client/api/request.js

```diff
@@ -27,22 +29,39 @@ const _request = ( url, data, method ) => {
 		request.body = JSON.stringify( data );
 	}
 
-	return fetch( baseURL + url, request ).then( ( response ) => {
-		return parseJson( response ).then( ( json ) => {
-			if ( json.success ) {
-				return json;
-			}
+	const poll = ( func ) => new Promise( ( resolve, reject ) => {
+		setTimeout( () => func().then( resolve ).catch( reject ), ASYNC_POLLING_INTERVAL );
+	} );
+
+	return ( function performRequest() {
+		return fetch( baseURL + url, request ).then( ( response ) => {
+			return parseJson( response ).then( ( json ) => {
+				if ( json.async_token ) {
+					request.body = JSON.stringify( { ...( data || {} ), async_token: json.async_token } );
+					return poll( performRequest );
+				}
+				switch ( json.async_status ) {
+					case 'IN_PROGRESS':
+						return poll( performRequest );
+					case 'NOT_FOUND':
+						throw { code: 'async_token_not_found', message: __( 'The request expired' ) }; // TODO: How to phrase this error?
```

Comment:
This is equivalent to a "503" error or something like that. It will happen if the server had the async request `IN_PROGRESS` for a few minutes and then it was purged from memory, so it can only be caused if the purchase failed in some fatal way, or for some reason the client hasn't polled for a few minutes.

I guess we could go with a generic `An error occured, please try again` (no need to refresh), what do you think?